core/src/Cache/Keys/MsalAccessTokenCacheKey.cs

```diff
@@ -0,0 +1,167 @@
+        private readonly string _clientId;
+        private readonly string _scopes;
+        private readonly string _tenantId;
+
+        internal MsalAccessTokenCacheKey(
+            string environment,
+            string tenantId,
+            string userIdentifier,
+            string clientId,
+            string scopes)
+        {
+            if (string.IsNullOrEmpty(environment))
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            if (string.IsNullOrEmpty(clientId))
+            {
+                throw new ArgumentNullException(nameof(clientId));
+            }
+
+           _environment = environment;
+           _homeAccountId = userIdentifier;
+           _clientId = clientId;
+           _scopes = scopes;
+           _tenantId = tenantId;
+        }
+
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append((_homeAccountId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(this._environment + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(MsalCacheConstants.AccessToken + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_clientId + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append((_tenantId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_scopes);
+
+            return stringBuilder.ToString();
+        }
+
+
+        #region UWP
+
+        /// <summary>
+        /// Gets a key that is smaller than 255 characters, which is a limitation for 
+        /// UWP storage. This is done by hashing the scopes and env.
+        /// </summary>
+        /// <remarks>
+        /// accountId - two guids plus separator - 73 chars        
+        /// "accesstoken" string - 11 chars
+        /// clientid - a guid - 36 chars
+        /// tenantid - a guid - 36 chars
+        /// scopes and env - a sha256 string - 44 chars
+        /// delimiters - 4 chars
+        /// total: 204 chars
+        /// </remarks>
+        public string GetFixedSizeKey()
+        {
+            var stringBuilder = new StringBuilder();
+            stringBuilder.Append((_homeAccountId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+            
+            stringBuilder.Append(MsalCacheConstants.AccessToken + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append(_clientId + MsalCacheConstants.CacheKeyDelimiter);
+            stringBuilder.Append((_tenantId ?? "") + MsalCacheConstants.CacheKeyDelimiter);
+
+            // can't use scopes and env because they are of variable length
+            stringBuilder.Append(CoreCryptographyHelpers.CreateSha256Hash(_scopes + _environment));
+
+            return stringBuilder.ToString();
+        }
+        #endregion
+
+
+        #region iOS
+
+        public string GetiOSAccountKey()
```

Comment:
That might be a bit difficult because it is used for other things than keys and I can't reference the TokenAccessor from here because it is compiled only on iOS.

---
In reply to: [221719634](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1287#discussion_r221719634) [](ancestors = 221719634)